### ISSUE TITLE
Why is the output format of shap_value of xgboost and RF different?

### ISSUE THREAD

**Issue by zevellong** [shap/shap#2675]
I use the exact same data, y has 2 classes, the output of **lightGBM and RF** are **lists**, and the list contains a two-dimensional matrix with the same dimension as the data, but the output of **xgboost** is **only a two-dimensional matrix**. 
, the xgboost plot and beeswam are the same

The `summary_plot` is also different, the xgboost plot is the same as `beeswam`.
``` python
# XGBoost: <class 'numpy.ndarray'> (32561, 12)
print("XGBoost:", type(shap_value), shap_value.shape)

# lightGBM: <class 'list'> (32561, 12) (32561, 12)
print("lightGBM:", type(shap_values1), shap_values1[0].shape, shap_values1[1].shape) 

# RF: <class 'list'> (32561, 12) (32561, 12)
print("RF:", type(shap_values2), shap_values2[0].shape, shap_values2[1].shape)
```

The test code is as follows:
``` python
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
import lightgbm as lgb
import shap
import numpy as np
import pandas as pd
import xgboost 

X,y = shap.datasets.adult()
X_display,y_display = shap.datasets.adult(display=True)
# create a train/test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=7)

# XGBoost
d_train = xgboost.DMatrix(X_train, label=y_train)
d_test = xgboost.DMatrix(X_test, label=y_test)
params = {
    "eta": 0.01,
    "objective": "binary:logistic",
    "subsample": 0.5,
    "base_score": np.mean(y_train),
    "eval_metric": "logloss"
}
model = xgboost.train(params, d_train, 5000, evals = [(d_test, "test")], verbose_eval=100, early_stopping_rounds=20)
explainer = shap.TreeExplainer(model)
shap_values = explainer.shap_values(X)
shap.summary_plot(shap_values, X_display, plot_type="bar")
print("XGBoost:", type(shap_values), shap_values.shape)

# lightGBM
l_train = lgb.Dataset(X_train, label=y_train)
l_test = lgb.Dataset(X_test, label=y_test)
params = {
    "max_bin": 512,
    "learning_rate": 0.05,
    "boosting_type": "gbdt",
    "objective": "binary",
    "metric": "binary_logloss",
    "num_leaves": 10,
    "verbose": -1,
    "min_data": 100,
    "boost_from_average": True
}

model1 = lgb.train(params, l_train, 10000, valid_sets=[l_test], early_stopping_rounds=50, verbose_eval=1000)
explainer1 = shap.TreeExplainer(model1)
shap_values1 = explainer1.shap_values(X)
shap.summary_plot(shap_values1, X)
print("lightGBM:", type(shap_values1), shap_values1[0].shape, shap_values1[1].shape) 

# RF
rf = RandomForestClassifier(max_features=6, max_depth=4).fit(X, y)
explainer2 = shap.TreeExplainer(rf)
shap_values2 = explainer2.shap_values(X)
print("RF:", type(shap_values2), shap_values2[0].shape, shap_values2[1].shape)



```

**Comment by zevellong** [comment]
I tried to trace the call, the call process  is as follows:
- **XGBoost**: `shap_values()` --> `phi=xgbooost.predict(X, pred_contribs=True)` --> `out=phi[:, :-1]` --> `return out`
- **lightGBM**:`shap_values()`-->`phi=lightgbm.predict(X, pred_contrib=True)`--> `phi2 = np.concatenate((0-phi, phi), axis=-1)` -->`out = [-phi[:,:-1], phi[:,-1]]`
- **RF**:  `shap_values()` --> ` phi = np.zeros((X.shape[0], X.shape[1]+1, num_outputs))` --> `_cext.dense_tree_shap(phi)` --> `out=[phi[:,:,0], phi[:,:,1]]` 

`pred_contribs=True` invokes the c++ version of the shap calculation, `out=phi[:, :-1]` removes the `expected_value` of the last column of $phi$. But in `lightGBM` calculation, a copy of $-phi$ needs to be copied after calculation.
Why does `lightGBM` make a copy of $-phi$ but `XGBoost` doesn't?
